### DISCUSSION TITLE
How to programmatically get initialization status?

### DISCUSSION THREAD

**Discussion by richin13** [neoclide/coc.nvim#4974] [Q&A] [answered]
Is there a way to know whether coc / coc-plugin is loading? I'm trying to write a custom component for my status line that shows a spinner is coc is loading but I can't find any info on how to get that information. If I use the default `coc#status` I see something like `⠦ Initializating pyright` (not exactly that cause it dissapears quite fast). Is there a function I can use to get whether something is initializing? Thanks!

**Comment by fannheyward** [comment, marked as answer]
Does `g:coc_service_initialized` work for you? Check `:h g:coc_service_initialized`, this var will be set to 1 after coc.nvim initialized.

**Reply by richin13** [reply]
Aha! That's exactly what I was looking for, not sure how I missed it 😅 🤦🏼 

Thank you!

**Comment by kazuya-hatsushi** [comment]
## 調査結果：useInitEffect の使用状況と必要性

Claude Code による調査を実施しました。

### 使用状況サマリー

`useInitEffect` およびその関連フックは、**主に共通コンポーネント・共通フックで使用**されています。

| フック名 | 使用箇所数 | 用途 |
|---------|-----------|------|
| `useInitEffect` | 7箇所 | 初期化時のイベントリスナー登録、DOM操作、フォーカス設定 |
| `useInitLayoutEffect` | 4箇所 | レイアウト計算、リサイズ監視、オフセット設定 |
| `useConstMemo` | 3箇所 | デフォルトpropsやフォームの初期値定義 |

### 詳細な使用箇所

#### 共通コンポーネント（elementsディレクトリ）
- **Toast.component.tsx**: 初期化時のアニメーション・タイマー設定
- **Modal.component.tsx**: デフォルトpropsの定義（`useConstMemo`）
- **PopupArea.component.tsx**: ポップアップの位置計算・フォーカス設定
- **SearchForm.component.tsx**: フォーカス設定
- **table.hook.ts**: スクロール同期、リサイズ監視、オフセット設定
- **MultiRowTable.hook.ts**: スクロール同期、リサイズ監視、オフセット設定

#### 共通フック
- **useRestoreForm.hook.ts**: フォーム復元の初期化処理

#### 機能コンポーネント
- **OrganizationForm.component.tsx**: フォームのデフォルト値（`useConstMemo`）

#### ストーリーファイル
- **SearchForm.stories.tsx**: ストーリー用のデフォルト値（`useConstMemo`）

### 現状の評価

#### メリット
- 意図が明確：「初期化時のみ実行」という意図がコードから読み取りやすい
- 共通コンポーネントで広く使用されている

#### デメリット
- **本来の存在意義が薄れている**：ESLintルール `react-hooks/exhaustive-deps` が無効化されているため、ルール回避という目的が失われている
- 独自フックの維持コスト
- 一般的なReactパターンではない

### 削除した場合の影響

#### 置き換え作業
- `useInitEffect` → `useEffect(callback, [])` に置き換え（7箇所）
- `useInitLayoutEffect` → `useLayoutEffect(callback, [])` に置き換え（4箇所）
- `useConstMemo` → `useMemo(callback, [])` に置き換え（3箇所）
- **合計14箇所の変更が必要**

#### リスク
- 共通コンポーネント（Toast、Modal、Table、Popup等）への変更となるため、**広範囲に影響する可能性**
- 機能的な差異はないが、念のため全体的な動作確認が必要

### 推奨アクション

以下の2つの選択肢を提案します：

**選択肢A: 段階的な削除**
1. まず新規コードでは `useInitEffect` を使用しない（通常の `useEffect` を使用）
2. 既存コードは徐々に置き換え
3. 最終的にフックを削除

**選択肢B: 現状維持＋ESLintルール有効化**
- `exhaustive-deps` ルールを有効化し、`useInitEffect` の存在意義を復活させる
- ただし、依存配列の誤りが多数検出される可能性あり

個人的には、**選択肢Aの段階的な削除**を推奨します。理由：
- Reactの標準的なパターンに戻すことで、新規メンバーの学習コストを削減
- ESLintルールの有効化は別途検討可能（`useInitEffect` の有無とは独立した判断）

---

**実施した調査コマンド**
```bash
# useInitEffect の実装確認
cat src/hooks/useInitEffect.hook.ts

# 使用箇所の検索
grep -r "useInitEffect\|useInitLayoutEffect\|useConstMemo" src --include="*.ts" --include="*.tsx"

# ESLint設定の確認
cat eslint.config.cjs | grep -A 2 -B 2 exhaustive-deps
```
